### PR TITLE
Minor improvements

### PR DESCRIPTION
**Pull request by Marwes**
Spotted some minor improvements while reading through #40 .

### CODE REVIEW

File: src/classes.rs [hunk 1]
```diff
@@ -149,14 +149,21 @@ impl<'a> Iterator for ByteClassRepresentatives<'a> {
...
+            .finish()
+    }
+}
 
 impl ByteClassBuilder {
     /// Create a new builder of byte classes where all bytes are part of the
     /// same equivalence class.
     pub fn new() -> ByteClassBuilder {
-        ByteClassBuilder(vec![false; 256])
+        ByteClassBuilder([false; 256])
```
Comment (BurntSushi) [inline src/classes.rs]: Is there a particular motivation for this? It seems like any performance overhead of the additional allocation would be negligible compared to the additional work being done to build an AC automaton.

File: src/prefilter.rs [hunk 2]
```diff
@@ -125,15 +125,30 @@ impl PrefilterState {
...
+            .field("byteset", &&self.byteset[..])
+            .finish()
+    }
 }
 
 impl StartBytesBuilder {
     /// Create a new builder for constructing a start byte prefilter.
     pub fn new() -> StartBytesBuilder {
-        StartBytesBuilder { byteset: vec![false; 256] }
+        StartBytesBuilder { byteset: [false; 256] }
```
Comment (BurntSushi) [inline src/prefilter.rs]: Same question as for the equivalence classes. Why? Doing this now means we need to maintain our own `Clone` and `Debug` impls.

File: src/nfa.rs [hunk 3]
```diff
@@ -322,6 +323,42 @@ impl<S: StateID> State<S> {
     }
 }
 
+#[derive(Clone, Debug)]
+struct Dense<S>(Vec<S>);
+
+impl<S> Dense<S>
+    where S: StateID,
```
Comment (BurntSushi) [inline src/nfa.rs]: I prefer keeping the style consistent with the surround code. Could you write `impl<S: StateID> Dense<S>` instead? (I generally only use `where` clauses when the bounds get more complicated.)

File: src/nfa.rs [hunk 4]
```diff
@@ -322,6 +323,42 @@ impl<S: StateID> State<S> {
...
+}
+
+impl<S> IndexMut<u8> for Dense<S> {
+    #[inline(always)]
+    fn index_mut(&mut self, i: u8) -> &mut S {
+        // SAFETY: This is safe because all dense transitions have
+        // exactly 256 elements, so all u8 values are valid indices.
+        unsafe { self.0.get_unchecked_mut(i as usize) }
+    }
+}
```
Comment (BurntSushi) [inline src/nfa.rs]: Ah this is great! (I am also kind of curious whether using `unsafe` here is necessary at all. It feels like a bounds check that LLVM should be able to elide.)

File: src/classes.rs (hunk 1 above)
Comment (Marwes) [reply src/classes.rs]: Yeah it doesn't have enough of an effect to be noticeable. Just figured that the added impls wouldn't need to be touched again after so it wouldn't be any burden. I can remove the change if you want.

Alternatively it the `[S; 256]` might be used enough to warrant an abstraction which can be used everywhere in the crate. It does incur a rather unfortunate `Clone` impl however... :/ https://github.com/Marwes/aho-corasick/commit/322e19b853069b28b4392feb631cc8a14e5eb1a2

File: src/nfa.rs (hunk 3 above)
Comment (Marwes) [reply src/nfa.rs]: Fixed

File: src/nfa.rs (hunk 4 above)
Comment (Marwes) [reply src/nfa.rs]: Yeah, it is unlikely to have an effect after LLVM is done with the code. It is one less thing for it to optimize I guess 🤷‍♂️

File: src/classes.rs (hunk 1 above)
Comment (BurntSushi) [reply src/classes.rs]: Yeah I'd probably prefer just switching back to `Vec<bool>`. I definitely don't think the generic structure is worth it.

File: src/nfa.rs (hunk 4 above)
Comment (BurntSushi) [reply src/nfa.rs]: Aye, next time I think of it I'll check out the before/after benchmarks with the unsafe removed here. If I can't see a difference, then I'll drop it.

File: src/classes.rs (hunk 1 above)
Comment (Marwes) [reply src/classes.rs]: 👍 Removed those commits.

(I can remove the unsafe with a macro and copy paste https://github.com/Marwes/aho-corasick/commit/87c3602845b61f3223a9c56c00c4fc80831f7679 if that makes `ByteMap` more palatable).

Comment (Marwes) [reply src/classes.rs]: Hmm, I guess the `[S; 256]` constraint is mostly self-inflicted. Only `ByteClasses` used an array beforehand so `ByteMap` could just use a `Vec` and `ByteClasses` could use an array still.

Comment (BurntSushi) [reply src/classes.rs]: I mostly just think that generics are overkill for this. :-) Sorry!

Comment (Marwes) [reply src/classes.rs]: Aww :( . Everything should be fixed in the PR though, the `ByteMap` stuff was all in a separate branch.

**Comment by BurntSushi** [comment]
I've squashed this PR down to two commits and rolled them into #45. Thanks!

**Pull request closed without merging**